st2common/st2common/hooks.py

```diff
@@ -37,14 +37,38 @@ def after(self, state):
 
         origin = state.request.headers.get('Origin')
         origins = cfg.CONF.api.allow_origin
+        serve_webui_files = cfg.CONF.api.serve_webui_files
+
+        # Build a list of the default allowed origins
+        api_port = cfg.CONF.api.port
+        public_api_url = cfg.CONF.auth.api_url
+        default_allowed_origins = []
+
+        # Default WebUI URL
+        default_allowed_origins.append('http://localhost:3000')
+
+        if serve_webui_files:
+            # Local API URL
+            default_allowed_origins.append('http://localhost:%s' % (api_port))
+            default_allowed_origins.append('http://127.0.0.1:%s' % (api_port))
+
+        if serve_webui_files and public_api_url:
+            # Public API URL
+            default_allowed_origins.append(public_api_url)
+
```

Comment:
Strictly speaking, you don't need this two IF's, the only thing you need to make sure is that url in your webui config exactly matches the url you are opening in your browser.

So, say, we have default devenv running in vm. Browser won't ask server for CORS if you open http://172.168.50.50:9101/webui in your browser and then pick Devenv from drop down (which also corresponds to http://172.168.50.50:9101). But if you for some reason would manage to run the browser on a VM itself and try to access web ui as http://localhost:9101/webui while still picking Devenv as an API server, you'll get an error because your Origin `http://localhost:9101` doesn't match API address `http://172.168.50.50:9101`. And it wont be a problem if you create another webui config record that points to http://localhost:9101 instead.

Yeah, CORS is kind of awful this way...